### ISSUE TITLE
Question about gradient_accumulation_steps logic in DDP mode

### ISSUE THREAD

**Issue by iminfine** [keyonvafa/inductive-bias-probes#12]
Hi,

I encountered a confusing behavior regarding `gradient_accumulation_steps` when training with multiple nodes/GPUs in DDP mode.

In `src/train_utils.py`, line 204:
https://github.com/keyonvafa/inductive-bias-probes/blob/main/inductivebiasprobes/src/train_utils.py#L201

```python
new_config["gradient_accumulation_steps"] //= ddp_world_size
```

And then later in `init_model` (line 433):
```python
effective_batch_size = (
    config["batch_size"]
    * config["gradient_accumulation_steps"]
    * (int(os.environ["WORLD_SIZE"]) if ddp else 1)
)
```

When scaling up to a large number of GPUs (e.g., 32 GPUs), if I set `--gradient_accumulation_steps 16` (to achieve a specific global batch size), the integer division `16 // 32` results in `0`, causing a `ZeroDivisionError` later in `iters_per_epoch` calculation:

```python
iters_per_epoch = dataset_size // effective_batch_size  # effective_batch_size becomes 0
```

1. What is the intention behind dividing `gradient_accumulation_steps` by `ddp_world_size`? It seems to imply that the argument provided should be the **total** accumulation steps across all GPUs, rather than per-GPU accumulation steps.
2. If I want to maintain a constant effective batch size (e.g., 4096) while scaling from 8 GPUs to 32 GPUs, what is the recommended way to set `--batch_size` and `--gradient_accumulation_steps`?
   - With the current logic, it seems I would need to set `--gradient_accumulation_steps` to a very large number (e.g., 512 for 32 GPUs to get 16 steps per GPU), which feels counter-intuitive.
3. Could you please share the training configuration used for the pre-trained models? specifically:
   - Number of GPUs/Nodes
   - Batch Size per GPU `--batch_size`
   - Gradient Accumulation Steps `--gradient_accumulation_steps`
   - Total Effective Batch Size

Thanks!

**Comment by petergchang** [comment]
Hi,

1. As you've correctly deduced, `gradient_accumulation_steps` is the total number of accumulation steps across all GPUs, rather than per-GPU steps (which follows the convention of Andrej Karpathy's `nanogpt` (see [link](https://github.com/karpathy/nanoGPT/blob/3adf61e154c3fe3fca428ad6bc3818b27a3b8291/train.py#L94-L95))). Therefore, we assume that `gradient_accumulation_steps` >= `ddp_world_size` (number of GPUs), although you bring up a good point that the default behavior of raising an error otherwise may not be user-friendly.

2. To maintain a constant effective batch size while scaling up the number of GPUs, you actually do not have to modify the `--batch_size` and `--gradient_accumulation_steps`. Since the effective batch size is the total number of samples processed across all GPUs, it is simply computed as `batch_size * gradient_accumulation_steps`. For instance, if you set your `--batch_size 128` and `--gradient_accumulation_steps 32`, then the effective batch size remains 4,096 regardless of your number of GPUs.

3. For the main experiments, we used `--batch_size 64` and `--gradient_accumulation_steps 8` with 8 GPUs.

Thanks!
